### PR TITLE
TAR-196 - building a generator for webapp features

### PR DESCRIPTION
**Pull request by lukehansell-hx**
#### What does this PR do?
This project is a [Yeoman generator](yeoman.io) which will builds features for our [webapp](https://github.com/holidayextras/webapp-template) projects.

For more info on this see the Readme within this PR.

Basically it looks for a package.json file and assumes that is your project root. It then expects a code directory and the code directory to contain a package.json file. If these exist then it will prompt you for the feature name. It then creates the necessary directory structure, writes a basic controller and updates the *code/index.js* file to tie in the newly created feature.

#### What unit or integration tests does this PR have?
I've written unit tests for this which are included within this PR.

#### What selenium tests does this PR have?
None

#### How should a developer review this?
- Check through the code and make sure you can easily follow it and that what it's doing makes sense to you.
- run the tests with the `npm test` command.

#### How should this be manually tested?
- install yeoman `npm i -g yo`
- install this generator using npm from repo or use `npm link`
- clone the webapp-template repo `git clone git@github.com:holidayextras/webapp-template.git`
- cd into the webapp-template dir `cd webapp-template`
- edit `code/index.js` adding the following:
    - `//====== yeoman requireHook =====//` below the require call for the productController
    - `//===== yeoman controllerHook =====//` within the `this.controller` declaration in `initialize()`
- run the generator `yo` (and follow instructions)  -or- `yo webapp-template`

check that the file structure defined in the README is created:

```
path/to/code
├ [featureName]
│ ├ controllers
│ │ └ [featureName]Controller.js
│ ├ style
│ ├ view_controllers
│ └ views
```

check that the `code/index.js` file is updated with references to the new feature.

If you modify the created controller so that it has an actual route defined and then run `make init && make dev` in the webapp-template project you should be able to navigate to that route and see the controller in action. For instance change 

```
routes: {
  "": "index"
},
```

to 
```
routes: {
  "test": "index"
},
```

and then in your browser navigate to *localhost:5000/index.html#test* and you should see the action of the controller in the browser (default is to add a `<div class="[featureName]"></div>` to the document.body).

#### Any background context you want to provide?
If not already familiar you could read docs on:
- [Yeoman](yeoman.io)
- [webapp-template](https://github.com/holidayextras/webapp-template) (Read comments in files as well)

#### What are the relevant tickets?
[TAR-196](https://holidayextras.jira.com/browse/TAR-196)

#### Screenshots (if appropriate)
None

#### What gif best describes how you feel about this work?

---
#### Quality checklist (for PR author to complete BEFORE code review):
- [x] I've checked there is appropriate unit test coverage.
- [x] I've checked there is appropriate selenium test coverage.
- [x] I've included regression tests for any bug fixes.
- [x] I've updated documentation with any changes to procedures.
- [x] I've tested this on mobile Android, Safari, Chrome for any visual changes.
- [x] I've tested this on desktop IE9/10/11, Firefox, Chrome for any visual changes.
- [x] I've checked any significant visual changes with a member of the UX/UI pod.
- [x] I've checked this work against the requirements of the Jira.
- [x] I've added and updated translations for all supported languages
- [x] I've informed those who need to know of any unforeseen impact of this work (database migrations, deployment procedure changes).
- [x] I am ready for this to be code reviewed, merged and tested.
- [x] I've thought about writing a blog post about this

#### IP or Developer Review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] +1 from me!

#### Software Engineer or Developer review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I've checked this work against the requirements of the Jira.
- [x] +1 from me!

#### Software Engineer or project guru review:
- [x] I agree with the assumptions made in the quality checklist.
- [x] I’ve witnessed the work behaving as expected.
- [x] I’ve run all the tests locally and they pass.
- [x] I’ve checked for appropriate test coverage.
- [x] I’ve checked for coding anti-patterns.
- [x] I don’t believe this work introduces unnecessary technical debt.
- [x] +1 from me!

### CODE REVIEW

**Comment by hx-markterry** [comment]
could this include an empty main.less in the style directory?

**Comment by hx-markterry** [comment]
i think the directory "view_controllers" should be camel case as are the other files and dirs in this project

File: generators/app/index.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      done();
+    }.bind(this));
+  },
+  
+  configuringVariables: function(){
+    this.camelName = camelCase(this.name);
+    this.capitalizedName = capitalize(this.camelName);
+  },
+  
+  scaffoldDirectories: function(){
```
Comment (hx-markterry) [inline generators/app/index.js]:
```
['controllers', 'style', ....].forEach(function(dir){
  this.mkdir(this.camelName + '/' + dir);
});
```

**Comment by lukehansell-hx** [comment]
@terry 
- I've included a main.less in the style directory.
- the view_controllers directory naming is copied over from the webapp-template. I can modify it here and create a JIRA to modify the webapp-template if we think this is incorrect?

File: package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "generator-webapp-feature",
+  "version": "1.0.0",
+  "description": "A generator for WebApp features for use within HX WebApp projects",
+  "main": "index.js",
+  "scripts": {
+    "test": "mocha tests/**/*Test.js"
```
Comment (hx-markterry) [inline package.json]: can we add mocha to the dev dependancies, and use `node_modules/.bin/mocha` here so it does not need to be installed globally

File: generators/app/index.js (hunk 1 above)
Comment (lukehansell-hx) [inline generators/app/index.js]: I agree that this is repetitive, but I personally feel that for the sake of clarity the current method is more easily understood and doesn't add that much overall. I'm open to debate on the best practice here though.

Comment (hx-markterry) [inline generators/app/index.js]: happy to leave it as it is in this case

**Comment by hx-markterry** [comment]
Yup i think the dir naming should be consistent as it will be really hard to sort this later

**Comment by lukehansell-hx** [comment]
@terry I've standardised the file naming so that files and directories now use snake_case everywhere.

File: tests/generators/app/webapp_feature_generator_test.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    });
+    
+    describe('when required', function(){
+      var app;
+      
+      beforeEach(function(){
+        app = require('../../../generators/app');
+      });
+      
+      afterEach(function(){
```
Comment (hx-markterry) [inline tests/generators/app/webapp_feature_generator_test.js]: dont need this

**Comment by hx-markterry** [comment]
ok from me :ocean:

**Comment by Joezo** [comment]
i'll se/dev on this

**Comment by Joezo** [comment]
When I ran the generator I ended up with one camelCase and one snake_case controller. What's the expected behaviour here?
https://cloudup.com/c3prrpJyoDW

File: README.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+## About this generator
+---
+
+The [WebApp template](https://github.com/holidayextras/webapp-template) projects split different functionality into "features". 
+This generator provides a quick method to add a new feature to an app.
+
+For more information on how we define a feature see the [WebApp template documentation](https://github.com/holidayextras/webapp-template#what-constitutes-a-feature-anyway).
+
+This generator relies on your terminal's CWD to be in either the root or a child directory of a WebApp template project.
+Inside your code directory the a new directory will be created:
```
Comment (Joezo) [inline README.md]: 'the a new'

File: README.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    
+    for the require statement make sure the above is at the top of the file with the other calls to `require()`.
+- `//===== yeoman controllerHook =====//`
+
+    to automatically insert the controller into those used by the app ensure the above is written in the JSON object for the instantiation of `this.controllers = {}`.
+
+These comments are required for the [Yeoman Hooks](http://remy.bach.me.uk/blog/2013/10/updating-existing-files-with-yeoman/) to update the file.
+If they are not present the generator will not error, but you will have to manually update the `code/index.js` file.
+
+*If you use the latest version of the webapp-template as the base of your app then the hooks should already be provided.*
```
Comment (Joezo) [inline README.md]: Can we lose this whole section if that repo is going to be updated imminently?

Comment (lukehansell-hx) [inline README.md]: They're just here in case of pre-existing apps and apps in which they may have been accidentally removed.

Comment (Joezo) [inline README.md]: Ah okay, maybe we can pull this into a different part of the documentation? It feels like a lot to read to then learn that you didn't need too :)

File: README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+You will require some globally installed packages in order to effectively run the generators:
+
+`npm i -g yo`
+
+This will install the [Yeoman](http://yeoman.io/) tool.
+
+Following this you can then install this generator directly from GitHub:
+
+`npm i -g [repo]`
```
Comment (Joezo) [inline README.md]: not sure if others are aware but you can install from github repo's really easily with npm `npm i -g holidayextras/generator-webapp-feature`

**Comment by lukehansell-hx** [comment]
the productController already exists in the webapp-template repo and will be updated in a PR after this one

File: README.md (hunk 5 above)
Comment (lukehansell-hx) [inline README.md]: I could move this to a "troubleshooting" section at the bottom of the README?

File: generators/app/index.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+var findParentDir = require('find-parent-dir');
+var fs = require('fs');
+var generators = require('yeoman-generator');
+var path = require('path');
+var snakeCase = require('snake-case');
+
+module.exports = generators.Base.extend({
+  
+  initializing: function(){
+    var done = this.async();
```
Comment (Joezo) [inline generators/app/index.js]: This function doesn't actually appear to be async...

File: generators/app/templates/feature_controller.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  routes: {
+    "<%= name %>": "index"
+  },
+
+  initialize: function(options) {
+    this.instances = options.instances;
+  },
+
+  index: function() {
+    React.render(<div className="<%= name %>"></div>, document.body);
```
Comment (Joezo) [inline generators/app/templates/feature_controller.js]: why do we wrap all of the components in divs?

Comment (lukehansell-hx) [inline generators/app/templates/feature_controller.js]: we don't. The div is the component.

File: README.md (hunk 5 above)
Comment (Joezo) [inline README.md]: Yeah that might be a better place, I think the initial part of the readme should be about getting started simply :)

**Comment by Joezo** [comment]
Left one remaining question, when that's answered i'll +1 :)

File: generators/app/templates/feature_controller.js (hunk 8 above)
Comment (Joezo) [inline generators/app/templates/feature_controller.js]: good point, well made

**Comment by Joezo** [comment]
:+1:

**Comment by hxtomprice** [comment]
I'll take IP/DEV

**Comment by hxtomprice** [comment]
Code all looks good. Might just want to add some notes in the readme on how to run the app though.

+1

**Pull request merged**